jetty-core/jetty-websocket/jetty-websocket-jetty-tests/src/test/java/org/eclipse/jetty/websocket/tests/ExplicitDemandTest.java

```diff
@@ -74,6 +81,38 @@ public void onWebSocketFrame(Frame frame, Callback callback)
         }
     }
 
+    @WebSocket(autoDemand = false)
+    public static class OnOpenSocket implements Session.Listener
+    {
+        CountDownLatch onOpen = new CountDownLatch(1);
+        BlockingQueue<String> textMessages = new BlockingArrayQueue<>();
+        Session session;
+
+        @Override
+        public void onWebSocketOpen(Session session)
+        {
+            try
+            {
+                this.session = session;
+                session.demand();
+                onOpen.await();
+            }
+            catch (InterruptedException e)
+            {
+                throw new RuntimeException(e);
+            }
+        }
+
+        @Override
+        public void onWebSocketFrame(Frame frame, Callback callback)
+        {
+            if (frame.getOpCode() == OpCode.TEXT)
+                textMessages.add(BufferUtil.toString(frame.getPayload()));
+            callback.succeed();
+
```

Comment:
As not demanding for a close frame is probably an oft repeated mistake, then it is probably best to have tests for both demanding the close and not demanding for it.

Reply:
> Strictly speaking, this method should demand, otherwise CLOSE frames won't be received.

I do not think this is true, demanding here will cause a `ReadPendingException`. Because there is no message sink registered, the JettyWebSocketFrameHandler will demand automatically when receiving a text/binary/continuation frame.

Maybe after addressing issue https://github.com/jetty/jetty.project/issues/11307 we would need to demand in the onFrame, but that is for a different PR.